app/src/androidTest/java/com/android/unio/end2end/EndToEndTest.kt

```diff
@@ -13,14 +18,26 @@ import com.google.firebase.auth.auth
 import com.google.firebase.firestore.firestore
 import com.google.firebase.functions.functions
 import dagger.hilt.android.testing.HiltAndroidRule
+import junit.framework.TestCase.assertEquals
 import okhttp3.OkHttpClient
 import okhttp3.Request
+import org.junit.After
 import org.junit.Before
 import org.junit.Rule
 
 open class EndToEndTest : FirebaseEmulatorFunctions {
+  init {
+    assertEquals(
+        """There are still listeners attached to the Auth instance. Make sure to remove 
+           them between tests with Firebase.auth.unregisterAllAuthStateListeners().
+        """
+            .trimIndent(),
+        0,
+        Firebase.auth.currentAuthStateListenerCount())
+  }
```

Comment:
Is there no way of removing the remaining listeners in this `init` block, if they are present?

Reply:
Yes, that assertion is made to check that no listeners are currently attached to the auth instance. If there are any, it means that they might still hold state from previous tests, which is something we want to avoid.

However, removing auth listeners is already handled by `TearDown()`, so all you need to do is extend that class in your tests and this assertion should never fail.

Also note that the assertion is not called in a `@Before` method because at that point the new listeners from `MainActivity` have already been created so the assertion would always fail, but when asserted from `init` we ensure that we only check the listeners of the previous tests.